server/src/test/java/org/elasticsearch/search/basic/SearchRedStateIndexIT.java

```diff
@@ -95,6 +99,89 @@ public void testClusterDisallowPartialsWithRedState() throws Exception {
         assertThat(ex.getDetailedMessage(), containsString("Search rejected due to missing shard"));        
     }
 
+    public void testDisallowPartialsWithRedStateRecovering() throws Exception {
```

Comment:
I wonder if it would be simpler to add a unit test in `FetchSearchPhaseTests` and `AbstractSearchAsyncActionTests`? We just need a child query that throws a ShardNotAvailableException and the reason why the exception is thrown is not really important. I also don't get why you need a test that fails reasonably often ? Shouldn't it fail all the time ?